File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    public int solution(int k, int[][] d) {
        int n = d.length, ans = 0;
        int[] idx = new int[n];
        for (int i = 0; i < n; i++) idx[i] = i;

        do {
            int f = k, cnt = 0;
            for (int i : idx) {
                if (f >= d[i][0]) {
                    f -= d[i][1];
                    cnt++;
                } else break;
            }
            ans = Math.max(ans, cnt);
        } while (next(idx));
        return ans;
    }

    boolean next(int[] a) {
        int i = a.length - 2;
        while (i >= 0 && a[i] >= a[i + 1]) i--;
        if (i < 0) return false;
        int j = a.length - 1;
        while (a[j] <= a[i]) j--;
        swap(a, i, j);
        for (int l = i + 1, r = a.length - 1; l < r; l++, r--) swap(a, l, r);
        return true;
    }

    void swap(int[] a, int i, int j) {
        int t = a[i]; a[i] = a[j]; a[j] = t;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    const N = dungeons.length;
    let max = 0;

    function getP(k, visited, count) {
        if(max <= count) {
                max = count;
            }
        
//         if(Object.keys(visited).length === N) {
//             // 더 이상의 던전 탐색은 그만
//             return;
//         }

        // 0, 1 방문 시점
        for(let i = 0; i < N; i ++) {
            if(!visited[i]){
                
                if(k >= dungeons[i][0]) {
                    visited[i] = true;
                getP(k - dungeons[i][1], visited, count+1);
                   visited[i] = false;
                }
            }
        }
    }

    const visited={}
    getP(k, visited, 0);
    return max;
}